h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -271,6 +276,10 @@ private void setFinalState() {
     if (res != null && res._output != null) {
       res._output._job = _job;
       res._output.stopClock();
+//      res.unlock(_job == null ? null : _job._key, false); // last resort: dirty way to force unlock to be able to reacquire lock
+      res.write_lock(_job);
+      res.update(_job);
+      res.unlock(_job);
```

Comment:
@michalkurka I fund this very frustrating that by simply adding this lock-update-unlock after model completion, I got a lot of tests failing due to a large diversity in how algos implementations handle locks:
- no abstraction or lock-update-unlock at high level, for example around `init` or `computeImpl`.
- unlocks not systematically applied in a `finally` block, so that the model can remain locked if an exception is thrown.
- the laurels go to GLM, where the locking logic was different for CV models and for the main model...

it was supposed to be a small fix, and turns into a pretty risky change (relying fully on our test coverage).
What do you think about coming up with better encapsulation for this locking logic, at least for model training?

Reply:
> What do you think about coming up with better encapsulation for this locking logic, at least for model training?

That seems like a good idea.

> unlocks not systematically applied in a finally block, so that the model can remain locked if an exception is thrown.

People really struggle with this, another common issue is assuming that the lock will always be acquired and then always trying to unlock the model in `finally`.

Better to do smaller steps - fix GLM and separately improve the framework.

Reply:
@seb:  GLM with CV was not implemented using the conventional method.  I am not sure why Tomas did it his way though.  Thanks you for fixing it.